src/modules/submodules/consensus.py

```diff
@@ -82,82 +105,255 @@ def _get_member_info(self, blockstamp: BlockStamp) -> MemberInfo:
+    def _get_current_frame(self, blockstamp: BlockStamp) -> tuple[SlotNumber, SlotNumber]:
+        """
+        function getCurrentFrame() external view returns (
+            uint256 refSlot,
+            uint256 reportProcessingDeadlineSlot
+        );
+        """
         consensus_contract = self._get_consensus_contract(blockstamp)
         return consensus_contract.functions.getCurrentFrame().call(
             block_identifier=blockstamp.block_hash,
         )
 
+    @lru_cache(maxsize=1)
+    def _get_frame_config(self, blockstamp: BlockStamp) -> tuple[int, int, int]:
+        """
+        struct FrameConfig {
+            uint64 initialEpoch;
+            uint64 epochsPerFrame;
+            uint64 fastLaneLengthSlots;
+        }
+        """
+        consensus_contract = self._get_consensus_contract(blockstamp)
+        return consensus_contract.functions.getFrameConfig().call(
+            block_identifier=blockstamp.block_hash,
+        )
+
+    # ----- Calculation reference slot for report -----
+
     def get_blockstamp_for_report(self, blockstamp: BlockStamp) -> Optional[BlockStamp]:
+        """Get blockstamp that should be used to build and send report for current frame."""
         member_info = self._get_member_info(blockstamp)
 
-        if blockstamp.slot_number < member_info.member_ref_slot:
+        latest_blockstamp = self._get_latest_blockstamp()
+
+        if not self.is_contract_reportable(latest_blockstamp):
+            logger.info({'msg': 'Contract is not reportable.'})
+            return
+
+        if latest_blockstamp.slot_number < member_info.current_frame_ref_slot:
             logger.info({'msg': 'Reference slot is not yet finalized.'})
             return
 
-        # Maybe check head slot number?
-        if blockstamp.slot_number > member_info.deadline_slot:
+        if not member_info.is_fast_line:
+            if latest_blockstamp.slot_number + member_info.fast_line_length_slot < member_info.current_frame_ref_slot:
+                logger.info({'msg': f'Member is not in fast lane, so report will be postponed for [{member_info.fast_line_length_slot}] slots.'})
+                return
+
+        if latest_blockstamp.slot_number > member_info.deadline_slot:
             logger.info({'msg': 'Deadline missed.'})
             return
 
-        return self._get_first_non_missed_slot(blockstamp, member_info.current_ref_slot)
+        return self._get_first_non_missed_slot(blockstamp, member_info.current_frame_ref_slot)
 
     def _get_first_non_missed_slot(self, blockstamp: BlockStamp, slot: SlotNumber) -> BlockStamp:
-        _, epoch_per_frame = self._get_frame_config(blockstamp)
+        _, epoch_per_frame = self._get_current_frame(blockstamp)
 
         for i in range(slot, slot - epoch_per_frame * 32, -1):
             try:
                 root = self.w3.cc.get_block_root(i).root
-            except KeyError:
+            except NotOkResponse as e:
```

Comment:
If error != 404, it can be caused by a node issue. I suggest continuing the lookup only at 404 OR raising the error immediately if not

Reply:
fixed